### DISCUSSION TITLE
How long will it take to develop the version modified using WebAssembly (wasm)?

### DISCUSSION THREAD

**Discussion by pengwuke** [Kitware/trame#567] [General]
Thank you for the author's contribution. How long will it take to develop the version modified using WebAssembly (wasm)?

**Comment by jourdain** [comment]
Can you explain what you mean? Client only with Python/wasm, or VTK/wasm?

**Reply by pengwuke** [reply]
Thank you for your response. Currently, the remote rendering of trame may experience lag due to network transmission issues, and its local rendering lacks many VTK features. It was previously mentioned that using WASM to implement local rendering could address this. When will this be available?

**Reply by jourdain** [reply]
It is mostly [there](https://github.com/kitware/trame-vtklocal) but assume some work in progress.

**Reply by pengwuke** [reply]
Thank you very much.

**Comment by Fevola** [comment]
Hello, I would like to know if wasm is available now? I followed the Running examples to execute the code, the installation went smoothly, but after starting the trame, the browser always shows loading. I would like to know if there are any additional settings that need to be made if i want to use wasm?

By the way, due to some component issues, I tried to replace VtkLocalView with VtkRemoteView. My computer has a GPU but the rendering performance is pretty poor, far below the expected 25-30 fps. The only modification I made in the code was to change VtkLocalView to VtkRemoteView. Should my code add something to make it better?

Thank you for your answer!

**Reply by jourdain** [reply]
Which "Running examples" did you followed? Did you try to clear the cache? Do you have any error on the server or client?

For wasm you can find some example of it [here](https://github.com/Kitware/trame-vtklocal/tree/master/examples). The only constraint is that you need a VTK version that has a compatible wasm version that was built against. In other words `vtk==9.3.20240817.dev0`.

If you are not seeing 25-30 fps with remote, something else is up. Like wrong driver or wrong vtk version (osmesa)?

**Reply by Fevola** [reply]
Thank you for your help!

The "Running examples" I mentioned are from the readme document in the project's root directory (PS: It seems that the vtk version in running examples needs to be updated). 
I downloaded `vtk==9.3.20240817.dev0` and have also tried clearing the browser cache, but I am still experiencing a prolonged loading time. There are no error messages on the server side, but the following messages appear in the browser:
```
paraview/:1       Failed to load resource: the server responded with a status of 405 (Method Not Allowed) 
vtkWasmSceneManager.mjs:1       Failed to load resource: the server responded with a status of 404 (Not Found) 
index-859ec636.js:23       Uncaught (in promise) Event
```

Regarding the remote render issue, I used `pip install vtk` to install it so it should be the normal version. I will try to check the driver

Thanks in advance

**Reply by jourdain** [reply]
Still don't know which example you are talking about. A link to a file would be helpful along with your `pip list` output from your venv.

**Reply by Fevola** [reply]
I apologize for not explaining clearly. What I mentioned was [this](https://github.com/Kitware/trame-vtklocal/blob/master/README.rst#running-examples).  Anyway it's now  working properly, thank you!
